File: libs/python/stonecharts/charts/funnel.py
"""Funnel chart renderer: ChartSpec -> SVG string.

Centered trapezoid stack with value->width scaling. Does NOT use
render_cartesian — funnel is the declared substrate exception (no axes).
See charts/funnel/design.md for the full geometry contract.
"""

from __future__ import annotations

from ..spec import ChartSpec, Gradient
from ..util import esc, fmt_num
from ._cartesian import a11y_summary, gradient_def, pattern_def


def render_svg(spec: ChartSpec) -> str:
    W, H = spec.width, spec.height
    theme = spec.theme
    palette = theme.palette
    cid = esc(spec.id)

    a11y_attr = ""
    a11y_desc = ""
    if spec.a11y:
        _sum = esc(a11y_summary(spec, "Funnel"))
        a11y_attr = f' role="img" aria-label="{_sum}"'
        a11y_desc = f"<desc>{_sum}</desc>"

    m_top: float = 20
    if spec.title:
        m_top += 26
    if spec.subtitle:
        m_top += 18
    m_left: float = 22
    m_right: float = 22
    m_bottom: float = 28 + (18 if spec.legend else 0)
    if spec.layout and spec.layout.margin:
        m = spec.layout.margin
        if m.top is not None:
            m_top = m.top
        if m.left is not None:
            m_left = m.left
        if m.right is not None:
            m_right = m.right
        if m.bottom is not None:
            m_bottom = m.bottom

    plot_x, plot_y = m_left, m_top
    plot_w, plot_h = W - m_left - m_right, H - m_top - m_bottom

    s0 = spec.series[0] if spec.series else None
    data = list(s0.data) if s0 else []
    n = len(data)
    cats = spec.x_axis.categories or [str(i) for i in range(n)]

    defs_parts: list[str] = []
    color_by_point = True
    funnel_fill = ""
    if s0 is not None:
        if isinstance(s0.color, Gradient):
            gid = f"{cid}-grad-0"
            defs_parts.append(gradient_def(gid, s0.color))
            funnel_fill = f"url(#{gid})"
            color_by_point = False
        elif s0.color:
            funnel_fill = esc(s0.color)
            color_by_point = False
        if s0.pattern is not None:
            pid = f"{cid}-pat-0"
            defs_parts.append(pattern_def(pid, s0.pattern))
            funnel_fill = f"url(#{pid})"
            color_by_point = False

    solid0 = ""
    if s0 is not None:
        if isinstance(s0.color, Gradient):
            solid0 = esc(s0.color.stops[0].color) if s0.color.stops else esc(palette[0])
        elif s0.color:
            solid0 = esc(s0.color)
        else:
            solid0 = esc(palette[0])

    subtype = spec.subtype or "funnel"
    min_w_frac = spec.min_width
    neck_w_frac = spec.neck_width
    neck_h_frac = spec.neck_height

    max_val = max(data) if data else 0.0

    def wscale(v: float) -> float:
        if max_val <= 0:
            return 0.0
        return plot_w * v / max_val

    p: list[str] = []

    _font = 'font-family="Segoe UI, Helvetica, Arial, sans-serif"'
    if spec.responsive:
        p.append(
            f'<svg class="sc-chart"{a11y_attr} xmlns="http://www.w3.org/2000/svg" '
            f'viewBox="0 0 {W} {H}" preserveAspectRatio="xMidYMid meet" '
            f'width="100%" {_font}>'
        )
    else:
        p.append(
            f'<svg class="sc-chart"{a11y_attr} xmlns="http://www.w3.org/2000/svg" '
            f'width="{W}" height="{H}" viewBox="0 0 {W} {H}" {_font}>'
        )

    if a11y_desc:
        p.append(a11y_desc)

    if defs_parts:
        p.append("<defs>" + "".join(defs_parts) + "</defs>")

    if theme.background:
        p.append(f'<rect class="sc-bg" x="0" y="0" width="{W}" height="{H}" fill="{theme.background}"/>')

    ty = 26
    if spec.title:
        p.append(
            f'<text class="sc-title" x="{W / 2:.1f}" y="{ty}" text-anchor="middle" '
            f'font-size="17" font-weight="600" fill="{theme.title_color}">{esc(spec.title)}</text>'
        )
        ty += 20
    if spec.subtitle:
        p.append(
            f'<text class="sc-subtitle" x="{W / 2:.1f}" y="{ty}" text-anchor="middle" '
            f'font-size="12" fill="{theme.subtitle_color}">{esc(spec.subtitle)}</text>'
        )

    p.append(
        f'<line class="sc-crosshair" x1="0" y1="{plot_y:.1f}" x2="0" y2="{plot_y + plot_h:.1f}" '
        f'stroke="{theme.crosshair_color}" stroke-width="1" stroke-dasharray="4 3" style="display:none"/>'
    )

    if n > 0 and s0 is not None:
        band_h = plot_h / n
        cx = plot_x + plot_w / 2

        order = list(range(n))
        if subtype == "pyramid":
            order = list(reversed(order))

        p.append('<g class="sc-series" data-series="0">')

        for draw_idx, stage_idx in enumerate(order):
            v = data[stage_idx]
            w_top = wscale(v)
            w_top = max(w_top, min_w_frac * plot_w)

            if subtype == "neck":
                neck_y = plot_y + plot_h * (1 - neck_h_frac)
                neck_w = neck_w_frac * plot_w
                y_top = plot_y + band_h * draw_idx
                y_bot = plot_y + band_h * (draw_idx + 1)

                if draw_idx < n - 1:
                    next_stage = order[draw_idx + 1]
                    w_next = wscale(data[next_stage])
                    w_next = max(w_next, min_w_frac * plot_w)
                else:
                    w_next = 0.0

                if y_top >= neck_y:
                    w_top = neck_w
                    w_bot = neck_w
                elif y_bot <= neck_y:
                    taper_h = neck_y - plot_y
                    if taper_h <= 0:
                        t_top = 0.0
                        t_bot = 0.0
                    else:
                        t_top = (y_top - plot_y) / taper_h
                        t_bot = (y_bot - plot_y) / taper_h
                    w0 = wscale(data[order[0]])
                    w0 = max(w0, min_w_frac * plot_w)
                    w_top = w0 + (neck_w - w0) * t_top
                    w_bot = w0 + (neck_w - w0) * t_bot
                else:
                    taper_h = neck_y - plot_y
                    t_top = 0.0 if taper_h <= 0 else (y_top - plot_y) / taper_h
                    w0 = wscale(data[order[0]])
                    w0 = max(w0, min_w_frac * plot_w)
                    w_top = w0 + (neck_w - w0) * t_top
                    w_bot = neck_w
            else:
                y_top = plot_y + band_h * draw_idx
                y_bot = plot_y + band_h * (draw_idx + 1)

                if draw_idx < n - 1:
                    next_stage = order[draw_idx + 1]
                    w_bot = wscale(data[next_stage])
                    w_bot = max(w_bot, min_w_frac * plot_w)
                else:
                    w_bot = 0.0 if subtype == "pyramid" else w_top

            x_tl = cx - w_top / 2
            x_tr = cx + w_top / 2
            x_bl = cx - w_bot / 2
            x_br = cx + w_bot / 2
            band_cy = y_top + band_h / 2

            fill = esc(palette[stage_idx % len(palette)]) if color_by_point else funnel_fill

            cat = cats[stage_idx] if stage_idx < len(cats) else str(stage_idx)
            s_name = s0.name

            p.append(
                f'<polygon class="sc-slice sc-point" data-series="0" '
                f'data-series-name="{esc(s_name)}" data-x="{esc(cat)}" '
                f'data-y="{esc(fmt_num(v))}" data-color="{fill}" '
                f'data-r="3.5" data-r-hover="6" '
                f'cx="{cx:.1f}" cy="{band_cy:.1f}" '
                f'points="{x_tl:.1f},{y_top:.1f} {x_tr:.1f},{y_top:.1f} {x_br:.1f},{y_bot:.1f} {x_bl:.1f},{y_bot:.1f}" '
                f'fill="{fill}"/>'
            )

        p.append("</g>")

    if spec.legend and spec.series:
        gap = 22
        est = [len(s0.name) * 7 + 26] if s0 else []
        total = sum(est) + gap * (len(est) - 1) if est else 0
        lx = plot_x + (plot_w - total) / 2
        ly = H - 10
        p.append('<g class="sc-legend">')
        p.append('<g class="sc-legend-item" data-series="0">')
        p.append(f'<rect x="{lx:.1f}" y="{ly - 9:.1f}" width="14" height="4" rx="2" fill="{solid0}"/>')
        p.append(
            f'<text x="{lx + 20:.1f}" y="{ly - 2:.1f}" font-size="12" '
            f'fill="{theme.legend_text_color}">{esc(s0.name)}</text>'
        )
        p.append("</g>")
        p.append("</g>")

    p.append("</svg>")
    return "".join(p)

File: libs/go/funnel.go
package stonecharts

import (
	"fmt"
	"strconv"
	"strings"
	"unicode/utf8"
)

func renderFunnelSVG(spec *ChartSpec) string {
	W, H := int(spec.Width), int(spec.Height)
	theme := spec.theme
	if theme == nil {
		t := lightTheme()
		theme = &t
	}
	palette := theme.Palette
	cid := esc(spec.ID)

	a11yAttr := ""
	a11yDesc := ""
	if spec.a11yOn() {
		sum := esc(a11ySummary(spec, "Funnel"))
		a11yAttr = fmt.Sprintf(` role="img" aria-label="%s"`, sum)
		a11yDesc = fmt.Sprintf("<desc>%s</desc>", sum)
	}

	mTop := 20.0
	if spec.Title != "" {
		mTop += 26
	}
	if spec.Subtitle != "" {
		mTop += 18
	}
	mLeft := 22.0
	mRight := 22.0
	mBottom := 28.0
	if spec.legendOn() {
		mBottom += 18
	}
	if spec.Layout != nil && spec.Layout.Margin != nil {
		m := spec.Layout.Margin
		if m.Top != nil {
			mTop = *m.Top
		}
		if m.Left != nil {
			mLeft = *m.Left
		}
		if m.Right != nil {
			mRight = *m.Right
		}
		if m.Bottom != nil {
			mBottom = *m.Bottom
		}
	}

	plotX, plotY := mLeft, mTop
	plotW, plotH := float64(W)-mLeft-mRight, float64(H)-mTop-mBottom

	var s0 *Series
	if len(spec.Series) > 0 {
		s0 = &spec.Series[0]
	}
	var data []float64
	if s0 != nil {
		data = s0.Data
	}
	n := len(data)
	cats := spec.XAxis.Categories
	if len(cats) == 0 {
		cats = make([]string, n)
		for i := 0; i < n; i++ {
			cats[i] = strconv.Itoa(i)
		}
	}

	var defsParts []string
	colorByPoint := true
	funnelFill := ""
	if s0 != nil {
		grad, solidHex := s0.colorSpec()
		if grad != nil {
			gid := cid + "-grad-0"
			defsParts = append(defsParts, gradientDef(gid, grad))
			funnelFill = "url(#" + gid + ")"
			colorByPoint = false
		} else if solidHex != "" {
			funnelFill = esc(solidHex)
			colorByPoint = false
		}
		if s0.Pattern != nil {
			pid := cid + "-pat-0"
			defsParts = append(defsParts, patternDef(pid, s0.Pattern))
			funnelFill = "url(#" + pid + ")"
			colorByPoint = false
		}
	}

	solid0 := ""
	if s0 != nil {
		grad, solidHex := s0.colorSpec()
		if grad != nil {
			if len(grad.Stops) > 0 {
				solid0 = esc(grad.Stops[0].Color)
			} else {
				solid0 = esc(palette[0])
			}
		} else if solidHex != "" {
			solid0 = esc(solidHex)
		} else {
			solid0 = esc(palette[0])
		}
	}

	subtype := spec.Subtype
	if subtype == "" {
		subtype = "funnel"
	}
	minWFrac := fdef(spec.MinWidth, 0.0)
	neckWFrac := fdef(spec.NeckWidth, 0.3)
	neckHFrac := fdef(spec.NeckHeight, 0.25)

	maxVal := 0.0
	if len(data) > 0 {
		maxVal = data[0]
		for _, v := range data[1:] {
			if v > maxVal {
				maxVal = v
			}
		}
	}

	wscale := func(v float64) float64 {
		if maxVal <= 0 {
			return 0.0
		}
		return plotW * v / maxVal
	}

	var p strings.Builder

	font := `font-family="Segoe UI, Helvetica, Arial, sans-serif"`
	if spec.Responsive {
		p.WriteString(fmt.Sprintf(
			`<svg class="sc-chart"%s xmlns="http://www.w3.org/2000/svg" viewBox="0 0 %d %d" preserveAspectRatio="xMidYMid meet" width="100%%" %s>`,
			a11yAttr, W, H, font))
	} else {
		p.WriteString(fmt.Sprintf(
			`<svg class="sc-chart"%s xmlns="http://www.w3.org/2000/svg" width="%d" height="%d" viewBox="0 0 %d %d" %s>`,
			a11yAttr, W, H, W, H, font))
	}

	if a11yDesc != "" {
		p.WriteString(a11yDesc)
	}

	if len(defsParts) > 0 {
		p.WriteString("<defs>" + strings.Join(defsParts, "") + "</defs>")
	}

	if theme.Background != "" {
		p.WriteString(fmt.Sprintf(
			`<rect class="sc-bg" x="0" y="0" width="%d" height="%d" fill="%s"/>`,
			W, H, theme.Background))
	}

	ty := 26
	if spec.Title != "" {
		p.WriteString(fmt.Sprintf(
			`<text class="sc-title" x="%s" y="%d" text-anchor="middle" font-size="17" font-weight="600" fill="%s">%s</text>`,
			f1(float64(W)/2), ty, theme.TitleColor, esc(spec.Title)))
		ty += 20
	}
	if spec.Subtitle != "" {
		p.WriteString(fmt.Sprintf(
			`<text class="sc-subtitle" x="%s" y="%d" text-anchor="middle" font-size="12" fill="%s">%s</text>`,
			f1(float64(W)/2), ty, theme.SubtitleColor, esc(spec.Subtitle)))
	}

	p.WriteString(fmt.Sprintf(
		`<line class="sc-crosshair" x1="0" y1="%s" x2="0" y2="%s" stroke="%s" stroke-width="1" stroke-dasharray="4 3" style="display:none"/>`,
		f1(plotY), f1(plotY+plotH), theme.CrosshairColor))

	if n > 0 && s0 != nil {
		bandH := plotH / float64(n)
		cx := plotX + plotW/2

		order := make([]int, n)
		for i := 0; i < n; i++ {
			order[i] = i
		}
		if subtype == "pyramid" {
			for i, j := 0, n-1; i < j; i, j = i+1, j-1 {
				order[i], order[j] = order[j], order[i]
			}
		}

		p.WriteString(`<g class="sc-series" data-series="0">`)

		for drawIdx, stageIdx := range order {
			v := data[stageIdx]
			wTop := wscale(v)
			if minWFrac*plotW > wTop {
				wTop = minWFrac * plotW
			}

			var wBot float64
			var yTop, yBot float64

			if subtype == "neck" {
				neckY := plotY + plotH*(1-neckHFrac)
				neckW := neckWFrac * plotW
				yTop = plotY + bandH*float64(drawIdx)
				yBot = plotY + bandH*float64(drawIdx+1)

				if yTop >= neckY {
					wTop = neckW
					wBot = neckW
				} else if yBot <= neckY {
					taperH := neckY - plotY
					var tTop, tBot float64
					if taperH <= 0 {
						tTop = 0.0
						tBot = 0.0
					} else {
						tTop = (yTop - plotY) / taperH
						tBot = (yBot - plotY) / taperH
					}
					w0 := wscale(data[order[0]])
					if minWFrac*plotW > w0 {
						w0 = minWFrac * plotW
					}
					wTop = w0 + (neckW-w0)*tTop
					wBot = w0 + (neckW-w0)*tBot
				} else {
					taperH := neckY - plotY
					var tTop float64
					if taperH <= 0 {
						tTop = 0.0
					} else {
						tTop = (yTop - plotY) / taperH
					}
					w0 := wscale(data[order[0]])
					if minWFrac*plotW > w0 {
						w0 = minWFrac * plotW
					}
					wTop = w0 + (neckW-w0)*tTop
					wBot = neckW
				}
			} else {
				yTop = plotY + bandH*float64(drawIdx)
				yBot = plotY + bandH*float64(drawIdx+1)

				if drawIdx < n-1 {
					nextStage := order[drawIdx+1]
					wBot = wscale(data[nextStage])
					if minWFrac*plotW > wBot {
						wBot = minWFrac * plotW
					}
				} else {
					if subtype == "pyramid" {
						wBot = 0.0
					} else {
						wBot = wTop
					}
				}
			}

			xTL := cx - wTop/2
			xTR := cx + wTop/2
			xBL := cx - wBot/2
			xBR := cx + wBot/2
			bandCY := yTop + bandH/2

			var fill string
			if colorByPoint {
				fill = esc(palette[stageIdx%len(palette)])
			} else {
				fill = funnelFill
			}

			cat := strconv.Itoa(stageIdx)
			if stageIdx < len(cats) {
				cat = cats[stageIdx]
			}
			sName := s0.Name

			p.WriteString(fmt.Sprintf(
				`<polygon class="sc-slice sc-point" data-series="0" data-series-name="%s" data-x="%s" data-y="%s" data-color="%s" data-r="3.5" data-r-hover="6" cx="%s" cy="%s" points="%s,%s %s,%s %s,%s %s,%s" fill="%s"/>`,
				esc(sName), esc(cat), esc(fmtNum(v)), fill,
				f1(cx), f1(bandCY),
				f1(xTL), f1(yTop), f1(xTR), f1(yTop), f1(xBR), f1(yBot), f1(xBL), f1(yBot),
				fill))
		}

		p.WriteString("</g>")
	}

	if spec.legendOn() && len(spec.Series) > 0 {
		gap := 22.0
		var est []float64
		if s0 != nil {
			est = []float64{float64(utf8.RuneCountInString(s0.Name)*7 + 26)}
		}
		total := 0.0
		if len(est) > 0 {
			for _, e := range est {
				total += e
			}
			total += gap * float64(len(est)-1)
		}
		lx := plotX + (plotW-total)/2
		ly := float64(H - 10)
		p.WriteString(`<g class="sc-legend">`)
		p.WriteString(`<g class="sc-legend-item" data-series="0">`)
		p.WriteString(fmt.Sprintf(
			`<rect x="%s" y="%s" width="14" height="4" rx="2" fill="%s"/>`,
			f1(lx), f1(ly-9), solid0))
		p.WriteString(fmt.Sprintf(
			`<text x="%s" y="%s" font-size="12" fill="%s">%s</text>`,
			f1(lx+20), f1(ly-2), theme.LegendTextColor, esc(s0.Name)))
		p.WriteString("</g>")
		p.WriteString("</g>")
	}

	p.WriteString("</svg>")
	return p.String()
}
